sdk/python/feast/infra/gcp.py

```diff
@@ -68,14 +69,14 @@ def update_infra(
         self,
         project: str,
         tables_to_delete: Sequence[Union[FeatureTable, FeatureView]],
-        tables_to_keep: Sequence[Union[FeatureTable, FeatureView]],
+        tables_to_have: Sequence[Union[FeatureTable, FeatureView]],
         partial: bool,
     ):
         from google.cloud import datastore
 
         client = self._initialize_client()
 
-        for table in tables_to_keep:
+        for table in tables_to_have:
             key = client.key("Project", project, "Table", table.name)
```

Comment:
The `OnlineStore` and `OfflineStore` abstractions still seem off to me. It seems like we are splitting out the online store's reading/writing functionality, but we want to leave it's schema management inside the provisioner. This essentially splits the life cycle of the online store but retains a hard coupling between the components. Wouldn't it be cleaner to go all out? 

More specifically
1. Move the apply functionality into the OnlineStore and OfflineStore implementation classes (with read/write)
2. Let the `Provider` contain its own workflows like
```python
def apply(...):
   online_store.apply()
   # push_api.apply()
   offline_store.apply()
   # serving_api.apply()

def get_online_store()
   return self._online_store()
```
If we don't go this route, then it's not clear to me why we need a provisioner/provider. We can just work with the online store and offline store directly it seems.

Reply:
The assumption that the provisioner for some backends could need to provision more things than just tables for online serving, such as (maybe) iam roles and security groups and things of that nature.

Reply:
Yea, that's why I listed non-serving examples above (like the push_api), implying that the provisioner/provider would have custom logic somewhere.